packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,233 @@
+    this.eventBus_ = eventBusService
+  }
+
+  withTransaction(transactionManager: EntityManager): DiscountConditionService {
+    if (!transactionManager) {
+      return this
+    }
+
+    const cloned = new DiscountConditionService({
+      manager: transactionManager,
+      discountRuleRepository: this.discountRuleRepository_,
+      discountRepository: this.discountRepository_,
+      discountConditionRepository: this.discountConditionRepository_,
+      eventBusService: this.eventBus_,
+    })
+
+    cloned.transactionManager_ = transactionManager
+
+    return cloned
+  }
+
+  async retrieve(
+    conditionId: string,
+    config: FindConfig<DiscountCondition> = {}
+  ): Promise<DiscountCondition> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const conditionRepo = manager.getCustomRepository(
+        this.discountConditionRepository_
+      )
+
+      const query = this.buildQuery_({ id: conditionId }, config)
+      const condition = await conditionRepo.findOne(query)
+
+      if (!condition) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `DiscountCondition with id: ${conditionId} was not found`
+        )
+      }
+
+      return condition
+    })
+  }
+
+  protected resolveConditionType_(data: UpsertDiscountConditionInput):
+    | {
+        type: DiscountConditionType
+        resource_ids: string[]
+      }
+    | undefined {
+    switch (true) {
+      case !!data.products?.length:
+        return {
+          type: DiscountConditionType.PRODUCTS,
+          resource_ids: data.products!,
+        }
+      case !!data.product_collections?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_COLLECTIONS,
+          resource_ids: data.product_collections!,
+        }
+      case !!data.product_types?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TYPES,
+          resource_ids: data.product_types!,
+        }
+      case !!data.product_tags?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TAGS,
+          resource_ids: data.product_tags!,
+        }
+      case !!data.customer_groups?.length:
+        return {
+          type: DiscountConditionType.CUSTOMER_GROUPS,
+          resource_ids: data.customer_groups!,
+        }
+      default:
+        return undefined
+    }
+  }
+
+  async upsertCondition(
+    discountIdOrDiscount: string | Discount,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    let resolvedConditionType
+
+    return await this.atomicPhase_(
+      async (manager: EntityManager) => {
+        resolvedConditionType = this.resolveConditionType_(data)
+
+        if (!resolvedConditionType) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Missing one of products, collections, tags, types or customer groups in data`
+          )
+        }
+
+        const discountConditionRepo: DiscountConditionRepository =
```

Comment:
creating a variable should be the closest to where it is used. We should move that somewhere around line 179. There is no need to get the repo here